pkg/controllers/securitypolicy_controller.go

```diff
@@ -17,37 +21,105 @@ import (
-		log.Error(err, "unable to fetch object")
-		return ctrl.Result{}, err
+	log.Info("reconciling securitypolicy CR", "securitypolicy", req.NamespacedName)
+
+	if err := r.Client.Get(ctx, req.NamespacedName, obj); err != nil {
+		log.Error(err, "unable to fetch security policy")
+		return ctrl.Result{}, client.IgnoreNotFound(err)
 	}
 
-	log.Info("reconciling securitypolicy CR", "securitypolicy", req.NamespacedName)
+	if obj.ObjectMeta.DeletionTimestamp.IsZero() {
+		if err := r.Service.CreateOrUpdateSecurityPolicy(obj); err != nil {
+			log.Error(err, "failed to create or update security policy", "securitypolicy", req.NamespacedName)
+			return ctrl.Result{}, err
+		}
+		if !controllerutil.ContainsFinalizer(obj, finalizerName) {
+			controllerutil.AddFinalizer(obj, finalizerName)
+			if err := r.Client.Update(ctx, obj); err != nil {
+				return ctrl.Result{}, err
+			}
+		}
+	} else {
+		if containsString(obj.GetFinalizers(), finalizerName) {
+			if err := r.Service.DeleteSecurityPolicy(obj.UID); err != nil {
+				return ctrl.Result{}, err
+			}
 
-	if err := r.Service.CreateOrUpdateSecurityPolicy(obj); err != nil {
-		log.Error(err, "failed to create or update security policy", "securitypolicy", req.NamespacedName)
-		return ctrl.Result{}, err
+			controllerutil.RemoveFinalizer(obj, finalizerName)
+			if err := r.Client.Update(ctx, obj); err != nil {
+				return ctrl.Result{}, err
+			}
+		}
 	}
 
 	return ctrl.Result{}, nil
 }
 
-// SetupWithManager sets up the controller with the Manager.
-func (r *SecurityPolicyReconciler) SetupWithManager(mgr ctrl.Manager) error {
+func containsString(source []string, target string) bool {
+	for _, item := range source {
+		if item == target {
+			return true
+		}
+	}
+	return false
+}
+
+func (r *SecurityPolicyReconciler) setupWithManager(mgr ctrl.Manager) error {
 	return ctrl.NewControllerManagedBy(mgr).
 		For(&v1alpha1.SecurityPolicy{}).
 		Complete(r)
 }
+
+// Start setup manager and launch GC
+func (r *SecurityPolicyReconciler) Start(mgr ctrl.Manager) error {
+	err := r.setupWithManager(mgr)
+	if err != nil {
+		return err
+	}
+	go r.GarbageCollector()
+	return nil
+}
+
+func (r *SecurityPolicyReconciler) GarbageCollector() {
+	ctx := context.Background()
+	log.V(1).Info("garbage collector started")
+	for {
+		nsxPolicySet, err := r.Service.ListSecurityPolicy()
+		if err != nil {
+			log.Error(err, "failed to get security policy from local store")
+		}
+		policyList := &v1alpha1.SecurityPolicyList{}
+		err = r.Client.List(ctx, policyList)
+		if err != nil {
+			log.Error(err, "failed to list security policy CR")
+		} else {
+			CRPolicySet := sets.NewString()
+			for _, policy := range policyList.Items {
+				CRPolicySet.Insert(string(policy.UID))
+			}
+
+			for elem := range nsxPolicySet {
+				if CRPolicySet.Has(elem) {
+					continue
+				}
+				log.V(1).Info("GC collected SecurityPolicy CR", "UID", elem)
+				r.Service.DeleteSecurityPolicy(types.UID(elem))
+			}
+		}
+		time.Sleep(GCInterval)
```

Comment:
Maybe we can put sleep at the beginning of the loop, but may still be risky if store init takes a longer time